### ISSUE TITLE
Validating `Json[MyEnum]` field with `strict=True` throws `NotImplementedError`

### ISSUE THREAD

**Issue by andersk** [pydantic/pydantic#8311]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I would expect `model_validate(…, strict=True)` to always accept the output of `model_dump(…, round_trip=True)`. However, this fails for a model with a field of type `Json[MyEnum]` (given a subclass `MyEnum` of `enum.Enum`).

```pytb
Traceback (most recent call last):
  File "/tmp/test.py", line 15, in <module>
    MyModel.model_validate(d, strict=True)  # NotImplementedError
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/v/lib/python3.11/site-packages/pydantic/main.py", line 503, in model_validate
    return cls.__pydantic_validator__.validate_python(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
NotImplementedError: Cannot check isinstance when validating from json, use a JsonOrPython validator instead.
```

### Example Code

```Python
from enum import Enum
from pydantic import BaseModel, Json

class MyEnum(Enum):
    Foo = 0
    Bar = 1
    Baz = 2

class MyModel(BaseModel):
    e: Json[MyEnum]

m = MyModel.model_construct(e=MyEnum.Baz)
d = MyModel.model_dump(m, round_trip=True)
MyModel.model_validate(d)  # ok
MyModel.model_validate(d, strict=True)  # NotImplementedError
```

### Python, Pydantic & OS Version

```Text
             pydantic version: 2.5.2
        pydantic-core version: 2.14.5
          pydantic-core build: profile=release pgo=true
                 install path: /tmp/v/lib/python3.11/site-packages/pydantic
               python version: 3.11.6 (main, Oct  2 2023, 13:45:54) [GCC 12.3.0]
                     platform: Linux-6.6.2-x86_64-with-glibc2.38
             related packages: typing_extensions-4.8.0
```

**Comment by andersk** [comment]
I see the same failure with a field of type `Json[ipaddress.IPv4Address]` or `Json[pathlib.Path]`, but not with a field of type `Json[datetime.datetime]` or `Json[uuid.UUID]`.

**Comment by sydney-runkle** [comment]
@andersk,

Thanks for reporting this.

Based on the nature of the error, I think this might be related to https://github.com/pydantic/pydantic/issues/7720.

I'm adding this as a bug fix to our 2.6.0 milestone.

**Comment by Dreamsorcerer** [comment]
I get the same with a TypedDict.
